Implementation of CMA support

arporter opened the pull request:
This pull request completes issue #6.

arporter commented:
Eike has identified one further extension to the functionality added in this pull request. "Matrix-matrix" kernels will be permitted to have zero or more read-only scalar arguments. I'll add this change in now so please hold off reviewing...

arporter commented:
OK. I think I've satisfied Eike's new requirement. Over to Rupert for review...

arporter commented on doc/dynamo0p3.rst:
```diff
@@ -891,22 +894,34 @@ as the number of dofs for each of the dofmaps. The full set of rules is:
+           is ``nlayers``.
+
 
 Application/Inverse-Application
```
I think it is correct as this rule is for a kernel that applies a CMA operator which, by definition, has been assembled for a whole column - therefore we don't need explicit information on levels. The dofs we provide are (presumably) for the whole column. Does that make sense?

rupertford replied:
If the dofs are for the whole column then that makes sense but it is unclear from the text. If it could be made more clear that would be good.

arporter replied:
I thought that a dofmap is only ever for a single cell since we have contiguous data-layout in the vertical. i.e. kernels do `map(df)+k` to get the appropriate array index for dof `df` on level `k`. We don't pass in `nlevels` because we the CMA operator acts on whole columns. I've added a line to clarify.

rupertford replied:
I'm really sorry but I still don't understand the text. It says the field passed in is 1D. OK, so we treat the whole column as 1D as it is CMA cf. 2D in "standard" kernels. Fine. It then says the extent of this 1D array is the number of degrees of freedom. Well the degrees of freedom are for a single cell so according to the text the extent of the field passed in is the size of a single cell, not a whole column. Shouldn't the extent be the degrees of freedom for a column?

arporter replied:
Fields are always passed as 1D arrays with dimension `undf` - the unique number of dofs for the function space (See rule 3.2 in the Subroutine section: http://psyclone.readthedocs.io/en/stable/dynamo0p3.html#subroutine). This is not the same as 'ndf' which is the number of dofs *per cell*. I imagine that `undf` is the number of dofs in the local partition and we use the dofmap to pull out the ones relating to the current column/cell. I've updated the doc a little bit to emphasise that ndf is per-cell.

rupertford replied:
You are quite right. I was getting confused between undf and ndf.

arporter commented on doc/dynamo0p3.rst:
```diff
@@ -966,11 +981,13 @@ and ``ncell_3d`` scalar arguments. The full set of rules are then:
        1) Include it and its associated parameters (see Rule 5 of CMA
 	  Assembly kernels).
 
+    4) For each scalar argument specified in meta-data include the
```
Well spotted. The code is right, the documentation is wrong - I've corrected it (bc5376c).

arporter commented on examples/check_examples:
```diff
@@ -42,6 +42,11 @@ python ../../../src/generator.py -s ./omp_script.py alg.x90
 
+echo "Running example dynamo/eg7"
+cd dynamo/eg7
+python ../../../src/generator.py alg.x90
```
Good point. Feels good to do that :-) (70c93b1)

rupertford replied:
Looks good.

arporter commented on examples/dynamo/README:
```diff
@@ -7,9 +7,12 @@
+of PSyclone that is being illustrated).
 
 In order to use PSyclone you must first add it to your PYTHONPATH. The
 simplest way to do this is to change to the top-level directory of your
```
Done.

rupertford replied:
Great.

arporter commented on examples/dynamo/README:
```diff
@@ -57,22 +60,22 @@ algorithm layer.
     python ../../../src/generator.py -nodm -s ./omp_script.py alg.x90
 
-or with both MPI and OpenMP (and loop fusion)
+or code parallelised with both MPI and OpenMP (and loop fusion)
```
Done both (e8e65fd)

arporter commented on src/dynamo0p3.py:
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
-                    arg_types=["gh_operator", "gh_field", "gh_scalar"],
-                    arg_accesses=["gh_inc", "gh_write"])
-                if write_args:
+                    arg_types=["gh_real", "gh_integer"])
```
Good idea - done.

arporter commented on src/dynamo0p3.py:
```diff
@@ -873,6 +875,13 @@ def func_descriptors(self):
+    def cma_operation(self):
+        ''' Returns the type of CMA operation identified from the kernel
+        meta-data (one of 'assembly', 'apply' or 'matrix-matrix') or
+        "" if the kernel does not involve CMA operators '''
```
So it does. Fixed.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1188,6 +1200,17 @@ def __init__(self, schedule):
+        # in this one are a list where the first item is the
+        # corresponding CMA object and the second is whether the map
+        # corresponds to the 'to' or 'from' space.
+        self._unique_cbanded_maps = {}
```
I agree. I felt slightly uncomfortable when I did that but didn't pause to examine why. I like your suggestion so have changed it. 475e151.

rupertford replied:
Great. Looks better.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1200,6 +1223,40 @@ def __init__(self, schedule):
+                    cma_args = psyGen.args_filter(
+                        call.arguments.args,
+                        arg_types=["gh_columnwise_operator"])
+                    map_name = get_cbanded_map_name(
```
It wasn't too bad. Done this and the next one in 33a0c9e

rupertford replied:
Great.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1209,33 +1266,174 @@ def initialise_dofmaps(self, parent):
+                               entity_decls=decl_ind_map_names))
+
+
+class DynInvokeCMAOperators(object):
```
I think it is better the way it is - the new dofmap entities fit well within the existing `DynInvokeDofmaps` class and they are conceptually different to the CMA operators themselves.

rupertford replied:
OK.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1209,33 +1266,174 @@ def initialise_dofmaps(self, parent):
+
+    cma_same_fs_params = ["nrow", "bandwidth", "alpha",
+                          "beta", "gamma_m", "gamma_p"]
+    cma_diff_fs_params = ["nrow", "ncol", "bandwidth", "alpha",
```
This is the order that the parameters are passed in to the kernels so it matters. I've added a couple of comments.

rupertford replied:
OK, that makes sense.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1569,17 +1794,28 @@ def gen_code(self, parent):
                             entity_decls=op_proxy_decs))
+        cma_op_proxy_decs = self.unique_proxy_declarations(
+            "gh_columnwise_operator")
+        if len(cma_op_proxy_decs) > 0:
```
I'm not sure that was deliberate and your suggestion is more pythonic so I've changed my code (and resisted the temptation to change the others).

rupertford replied:
Your code is now officially better than its surrounding code.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1569,17 +1794,28 @@ def gen_code(self, parent):
         invoke_sub.add(CommentGen(invoke_sub, " Initialise number of layers"))
         invoke_sub.add(CommentGen(invoke_sub, ""))
 
         # Use the first argument that is not a scalar
```
Done.

arporter replied:
Oops, don't know what I did there. I've added a comment in the relevant location.

rupertford replied:
Great.

arporter commented on src/dynamo0p3.py:
```diff
@@ -1569,17 +1794,28 @@ def gen_code(self, parent):
+        cma_op = None
         for var in self.psy_unique_vars:
-            if var.type in ["gh_field", "gh_operator"]:
+            if not first_var and var.type in ["gh_field", "gh_operator",
```
Good idea - done.

arporter commented on src/dynamo0p3.py:
```diff
@@ -2345,6 +2605,13 @@ def _setup(self, ktype, module_name, args, parent):
     @property
+    def cma_operation(self):
+        ''' Returns the type of CMA operation performed by this kernel
+        (one of 'assembly', 'apply' or 'matrix-matrix') or None if the
```
Good spot. It returns None now (if kernel is not CMA-related).

arporter replied:
Belatedly fixed related test too.

arporter commented on src/dynamo0p3.py:
```diff
@@ -2619,11 +2886,24 @@ def generate(self):
+            self.mesh_height()
+        # Pass the number of cells in the mesh if this kernel has a
+        # LMA operator argument
+        # TODO this code should replace the code that currently includes
```
Another issue - we have a ticket (551) for it on SRS that I will migrate over to github.

arporter commented on src/dynamo0p3.py:
```diff
@@ -2849,9 +3193,11 @@ def scalar(self, scalar_arg):
-        ndf_name = get_fs_ndf_name(function_space)
-        self._arglist.append(ndf_name)
+        if self._kern.cma_operation not in ["matrix-matrix"]:
+            # There is currently one compulsory argument: "ndf" but only
```
Good point :-) I've renamed it `fs_common` and done this for the base class too - 487524a.

arporter commented on src/dynamo0p3.py:
```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
+        _local_args = [nrow]
+        if arg.function_space_to.orig_name != \
+           arg.function_space_from.orig_name:
+            # If the to- and from-spaces are the same then so are ncol and
```
I'm not quite sure what you mean. However, I've re-written the comment so that it is worded correctly for its location. By default we pass both `nrow` and `ncol` to a kernel but if the to- and from-spaces are the same then we only pass `nrow`.

rupertford replied:
Great.

arporter commented on src/dynamo0p3.py:
```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
+            raise GenerationError(
+                "Internal error: a CMA operator (gh_columnwise_operator) must "
+                "be supplied but got {0}".format(operator.type))
+        # If a kernel applies a CMA operator then it must only have a
```
That makes two of us. Must be a hangover from something I had during the development process. I've removed the comment.

arporter commented on src/dynamo0p3.py:
```diff
@@ -3053,6 +3421,83 @@ def operator(self, arg):
+                "be supplied but got {0}".format(operator.type))
+        # If a kernel applies a CMA operator then it must only have a
+        # single such operator amongst its arguments
+        if operator.function_space_to.orig_name == function_space.orig_name:
```
I've added a new comment to clarify.

rupertford replied:
Great.

arporter commented on src/dynamo0p3.py:
```diff
@@ -3597,11 +4044,20 @@ def get_arg_on_space(self, func_space):
-    def has_operator(self):
-        ''' Returns true if at least one of the arguments is an operator. '''
+    def has_operator(self, op_type=None):
+        ''' Returns true if at least one of the arguments is an operator
```
doc-string corrected.

arporter commented on src/dynamo0p3_builtins.py:
```diff
@@ -150,6 +150,11 @@ def qr_required(self):
+    def cma_operation(self):
+        ''' Built-ins do not perform operations with Column-Matrix-Assembly
+        operators '''
+        return ""
```
Good point. Fixed (return None).

arporter commented on src/tests/test_files/dynamo0p3/20.0_cma_assembly.f90:
```diff
@@ -0,0 +1,54 @@
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
No! Fixed.

arporter commented on src/tests/test_files/dynamo0p3/20.1.1_cma_apply_same_spaces.f90:
```diff
@@ -0,0 +1,55 @@
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
No :-) Fixed.

arporter commented on src/tests/test_files/dynamo0p3/20.1_cma_apply.f90:
```diff
@@ -0,0 +1,54 @@
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
Nope.

arporter commented on src/tests/test_files/dynamo0p3/20.3_cma_assembly_field.f90:
```diff
@@ -0,0 +1,53 @@
+
+program single_invoke_cma
+
+  ! Description: single function specified in an invoke call using all
```
Fixed.

arporter commented on src/tests/test_files/dynamo0p3/columnwise_op_app_same_fs_kernel_mod.F90:
```diff
@@ -0,0 +1,129 @@
+  !> @param [in] gamma_m banded matrix parameter \f$\gamma_-\f$
+  !> @param [in] gamma_p banded matrix parameter \f$\gamma_+\f$
+  !> @param [in] indirection_dofmap_to indirection map for to-space
+  !> @param [in] indirection_dofmap_from indirection map for from-space
```
I've corrected the list. I like having them there as it explains what the various arguments are.

arporter commented:
@rupertford I think I've addressed everything - back for 2nd review...

rupertford commented on src/dynamo0p3.py:
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
+            # operator
+            write_args = psyGen.args_filter(
+                self._arg_descriptors,
+                arg_types=["gh_operator", "gh_field", "gh_real", "gh_integer"],
```
Change looks good.

rupertford commented on src/dynamo0p3.py:
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
+            write_args = psyGen.args_filter(
+                self._arg_descriptors,
+                arg_types=["gh_operator", "gh_field", "gh_real", "gh_integer"],
+                arg_accesses=["gh_inc", "gh_write"])
```
Change looks good.

rupertford commented on src/dynamo0p3.py:
```diff
@@ -796,69 +814,53 @@ def _identify_cma_op(self, cwise_ops):
+                                      [arg.type for arg in write_args]))
+            if len(cwise_ops) == 1:
+
+                # If this is an assembly kernel then we need at least one
```
Done. Good.

rupertford commented on src/dynamo0p3.py:
```diff
@@ -1188,6 +1200,17 @@ def __init__(self, schedule):
+        # is a list where the first item correponds to the CMA object
+        # and the second is whether the map corresponds to the 'to' or
+        # 'from' space.
+        self._unique_indirection_maps = {}
```
Done here too. Great.

arporter commented:
@rupertford I think I've addressed the remaining issues (it's hard to follow - maybe we need to review how we use pull-requests?):

1. I suddenly understood what you meant and have added a comment at 1844 in dynamo0p3.py
2. I've added a line in dynamo0p3.rst to clarify why nlevels is not required when applying a CMA operator
3. I've added a comment saying we choose to pass nrow if ncol==nrow (3471 of dynamo0p3.py).

arporter commented:
@rupertford  - I think I'm done. Realised I needed to correct the copyright notice in some of the columnwise test kernels so I've done that.
